# Cleaned Company Employee Dataset

**Task11** — Load, inspect, clean, verify, and export a messy employee dataset.

In [1]:
import pandas as pd

df = pd.read_csv("Day11_Messy_Company_Employee_Dataset.csv")
print("Shape:", df.shape)
display(df.head())
print("\nData types:")
display(df.dtypes)
print("\nMissing values:")
display(df.isnull().sum())
print("\nDuplicate rows:", df.duplicated().sum())

Shape: (157, 12)


,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid



Data types:


,0
Employee_ID,object
Employee_Name,object
Department,object
Job_Title,object
Age,float64
Gender,object
Annual_Salary,float64
Experience_Years,float64
Joining_Date,object
City,object



Missing values:


,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,4
Gender,5
Annual_Salary,5
Experience_Years,3
Joining_Date,0
City,5



Duplicate rows: 7


## Cleaning approach

- Strip whitespace from text fields.
- Standardize known categorical variants.
- Use median imputation for missing numeric values.
- Use mode imputation for missing categorical values.
- Convert the joining date to datetime.
- Remove exact duplicate records with `drop_duplicates()`.

`dropna()` and forward filling are not used because dropping valid employee rows would lose information, while the dataset is not a time-series where forward filling is appropriate.

In [2]:
# Clean whitespace and categorical inconsistencies
for col in df.select_dtypes(include="object").columns:
    df[col] = df[col].apply(lambda x: x.strip() if isinstance(x, str) else x)

category_maps = {
    "Department": {"Human Resources": "HR", "Information Technology": "IT", "Fin": "Finance"},
    "Gender": {"M": "Male", "m": "Male", "F": "Female", "f": "Female"},
    "Work_Mode": {"WFO": "Office", "On-site": "Office", "On Site": "Office", "WFH": "Remote"}
}
for col, mapping in category_maps.items():
    df[col] = df[col].replace(mapping)

# Numeric conversion + median imputation
numeric_cols = ["Age", "Annual_Salary", "Experience_Years", "Performance_Score"]
for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors="coerce")
    df[col] = df[col].fillna(df[col].median())

# Categorical mode imputation
for col in ["Department", "Gender", "City", "Work_Mode"]:
    df[col] = df[col].fillna(df[col].mode()[0])

# Date conversion
df["Joining_Date"] = pd.to_datetime(df["Joining_Date"], errors="coerce")

# Remove duplicates
df = df.drop_duplicates().reset_index(drop=True)

display(df.head())

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,Marketing,Data Scientist,48.0,Other,108371.0,13.20,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.10,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,HR,HR Manager,28.0,Female,119400.0,13.20,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,9.35,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.90,2018-07-24,Hyderabad,4.0,Hybrid


In [3]:
print("Final shape:", df.shape)
print("\nFinal missing values:")
display(df.isnull().sum())
print("\nFinal duplicate rows:", df.duplicated().sum())
print("\nFinal data types:")
display(df.dtypes)

df.to_csv("Cleaned_Company_Employee.csv", index=False)
print("\nSaved: Cleaned_Company_Employee.csv")

Final shape: (150, 12)

Final missing values:


,0
Employee_ID,0
Employee_Name,0
Department,0
Job_Title,0
Age,0
Gender,0
Annual_Salary,0
Experience_Years,0
Joining_Date,0
City,0



Final duplicate rows: 0

Final data types:


,0
Employee_ID,object
Employee_Name,object
Department,object
Job_Title,object
Age,float64
Gender,object
Annual_Salary,float64
Experience_Years,float64
Joining_Date,datetime64[ns]
City,object



Saved: Cleaned_Company_Employee.csv


## Before vs After

| Metric | Before | After |
|---|---:|---:|
| Rows | 157 | 150 |
| Columns | 12 | 12 |
| Missing cells | 32 | 0 |
| Duplicate rows | 7 | 0 |

The cleaned CSV is exported after the final verification.

In [4]:
from google.colab import files

files.download("Cleaned_Company_Employee.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>